{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from sklearn.model_selection import train_test_split, StratifiedKFold, GridSearchCV, RandomizedSearchCV, cross_val_score\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.dummy import DummyClassifier\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.metrics import classification_report, confusion_matrix, accuracy_score\n",
    "from sklearn.impute import SimpleImputer\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.feature_selection import SelectFromModel\n",
    "from sklearn.inspection import permutation_importance\n",
    "from xgboost import XGBClassifier\n",
    "from imblearn.over_sampling import SMOTE\n",
    "from imblearn.pipeline import Pipeline as ImbPipeline\n",
    "import shap\n",
    "from fastshap import KernelExplainer\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data\n",
    "path = '/trinity/home/atheodosiadou/macro/raw_data/ChIP_comb/ChIP_comb_merged_per_condition/bed_intersect/H3K27ac_EGR1_LPS/H3K27ac_EGR1_LPS_ctrl_overlap_modified_008.bed'\n",
    "data = pd.read_csv(path, sep=\"\\t\", header=None)\n",
    "\n",
    "feature_names = [\"chr_h3k\", \"start_h3k\", \"end_h3k\", \"mvalue_h3k\", \"avalue_h3k\", \"pvalue_h3k\", \"peak_group_h3k\",\n",
    "                 \"normalized_density_1_h3k\", \"normalized_density_2_h3k\", \n",
    "                 \"chr_tf\", \"start_tf\", \"end_tf\", \"mvalue_tf\",\"avalue_tf\", \"pvalue_tf\", \"peak_group_tf\",\n",
    "                 \"normalized_density_1_tf\", \"normalized_density_2_tf\",\n",
    "                 \"annotation_gene\", \"distance_to_TSS\", \"symbol\", \"gene_name\", \"overlap\"]\n",
    "data.columns = feature_names\n",
    "\n",
    "fibrosis_genes_df = pd.read_csv('/trinity/home/atheodosiadou/macro/raw_data/fibrosis/uniprotkb_fibrosis_AND_model_organism_9_2025_06_19.tsv', sep=\"\\t\", header = 0)\n",
    "fibrosis_genes = fibrosis_genes_df['Gene Names'].str.split().explode().unique()\n",
    "print(fibrosis_genes)\n",
    "data['fibrosis'] = data['symbol'].isin(fibrosis_genes).astype(int)\n",
    "print(data)\n",
    "print(data['fibrosis'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Chromosome conversion\n",
    "\n",
    "def extract_chromosome_number(chrom):\n",
    "    try:\n",
    "        return int(chrom[3:])\n",
    "    except ValueError:\n",
    "        return {'X': 23, 'Y': 24}.get(chrom[3:], np.nan)\n",
    "\n",
    "data[\"chr_h3k\"] = data[\"chr_h3k\"].apply(extract_chromosome_number)\n",
    "data[\"chr_tf\"] = data[\"chr_tf\"].apply(extract_chromosome_number)\n",
    "\n",
    "# Label mappings\n",
    "h3k_mapping = {'H3K27ac_LPS.macs3.broad_peaks_filtered_unique': 0, 'H3K27ac_ctrl.macs3.broad_peaks_filtered_unique': 1, 'merged_common' : 2}\n",
    "tf_mapping = {'EGR1_LPS.macs3_peaks_filtered_unique': 0, 'EGR1_ctrl.macs3_peaks_filtered_unique': 1, 'merged_common': 2, '.': -1}\n",
    "\n",
    "def annotation_encoder(val):\n",
    "    if pd.isna(val) or val == \".\": return -1\n",
    "    val = val.lower()\n",
    "    return {\n",
    "        \"promoter_(<=1kb)\": 0, \"promoter_(1-2kb)\": 1, \"promoter_(2-3kb)\": 2,\n",
    "        \"intron\": 3, \"exon\": 4, \"5'utr\": 5, \"3'utr\": 6,\n",
    "        \"distal_intergenic\": 7, \"downstream\": 8\n",
    "    }.get(val, -1)\n",
    "\n",
    "# Feature prep\n",
    "X = data.drop([\"mvalue_h3k\", \"avalue_h3k\", \"pvalue_h3k\", \"normalized_density_1_h3k\", \"normalized_density_2_h3k\",\n",
    "               \"chr_tf\", \"start_tf\", \"end_tf\", \"avalue_tf\", \"pvalue_tf\", \"normalized_density_1_tf\", \"normalized_density_2_tf\",\n",
    "               \"distance_to_TSS\", \"symbol\", \"gene_name\"], axis=1)\n",
    "X['peak_group_h3k'] = X['peak_group_h3k'].map(h3k_mapping)\n",
    "X['peak_group_tf'] = X['peak_group_tf'].map(tf_mapping)\n",
    "X['annotation_gene'] = X['annotation_gene'].apply(annotation_encoder)\n",
    "X['overlap'] = X['overlap'].apply(lambda x: 1 if int(x) > 0 else 0)\n",
    "y = data[\"mvalue_h3k\"]\n",
    "print(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #keep only the 2 minority classes(mvalue_h3k == 1 and mvalue_h3k == -1)\n",
    "# data = data[(data['mvalue_h3k'] == 1) | (data['mvalue_h3k'] == -1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split before SMOTE\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.2, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify columns with \".\" and replace with NaN\n",
    "for col in X.columns:\n",
    "    try:\n",
    "        X_train[col] = pd.to_numeric(X_train[col])  # Try to convert to numeric\n",
    "        X_test[col] = pd.to_numeric(X_test[col])\n",
    "    except ValueError:\n",
    "        print(f\"Column {col} contains non-numeric values. Replacing '.' with NaN.\")\n",
    "        X_train[col] = X_train[col].replace('.', np.nan)  # Replace \".\" with NaN\n",
    "        X_test[col] = X_test[col].replace('.', np.nan)\n",
    "        X_train[col] = pd.to_numeric(X_train[col])  # Convert to numeric after replacement\n",
    "        X_test[col] = pd.to_numeric(X_test[col])\n",
    "\n",
    "# Impute missing values\n",
    "imputer = SimpleImputer(strategy='mean')\n",
    "X_train = imputer.fit_transform(X_train)\n",
    "X_test = imputer.transform(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stratified K-Fold\n",
    "skf = StratifiedKFold(n_splits=3, shuffle=True, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dummy classifier\n",
    "dummy = DummyClassifier(strategy='most_frequent')\n",
    "dummy_scores = cross_val_score(dummy, X_train, y_train, cv=skf, scoring='f1_weighted')\n",
    "print(\"\\nBaseline Dummy Classifier F1 (Weighted) CV Scores:\", dummy_scores)\n",
    "print(\"Mean F1:\", np.mean(dummy_scores))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Random Forest with SMOTE pipeline\n",
    "rf_pipeline = ImbPipeline([\n",
    "    ('smote', SMOTE(random_state=42)),\n",
    "    ('clf', RandomForestClassifier(class_weight='balanced', random_state=42))\n",
    "])\n",
    "param_grid_rf = {\n",
    "    'clf__n_estimators': [500, 1000],\n",
    "    'clf__max_features': ['sqrt', 'log2'],\n",
    "    'clf__max_depth': [None, 10, 20],\n",
    "    'clf__criterion': ['gini', 'entropy']\n",
    "}\n",
    "rf_search = RandomizedSearchCV(rf_pipeline, param_grid_rf, n_iter=5, scoring='f1_weighted', n_jobs=-1, cv=skf)\n",
    "rf_search.fit(X_train, y_train)\n",
    "print(\"\\nBest RF params:\", rf_search.best_params_)\n",
    "y_pred_rf = rf_search.best_estimator_.predict(X_test)\n",
    "print(\"\\nRandom Forest Classification Report:\")\n",
    "report = classification_report(y_test, y_pred_rf)\n",
    "print(report)\n",
    "\n",
    "# Save to a text file\n",
    "with open(\"/trinity/home/atheodosiadou/macro/results/ChIP_comb/models/fibrosis/H3K27ac_EGR1_LPS/classification_report_rf_merged_manorm_008.txt\", \"w\") as f:\n",
    "    f.write(report)\n",
    "\n",
    "\n",
    "\n",
    "# ])\n",
    "# param_grid = {\n",
    "#     'clf__n_estimators': [500, 1000],\n",
    "#     'clf__max_features': ['sqrt', 'log2'],\n",
    "#     'clf__max_depth': [None, 10, 20],\n",
    "#     'clf__criterion': ['gini', 'entropy'],\n",
    "#     'clf__min_samples_leaf': [1, 2, 4]\n",
    "# }\n",
    "# grid = GridSearchCV(rf_pipeline, param_grid, cv=skf, scoring='f1_weighted', n_jobs=-1)\n",
    "# grid.fit(X_train, y_train)\n",
    "# print(\"\\nBest RF params:\", grid.best_params_)\n",
    "# y_pred_rf = grid.best_estimator_.predict(X_test)\n",
    "# print(\"\\nRandom Forest Classification Report:\")\n",
    "# print(classification_report(y_test, y_pred_rf))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Logistic Regression pipeline\n",
    "lr_pipeline = ImbPipeline([\n",
    "    ('smote', SMOTE(random_state=42)),\n",
    "    ('clf', LogisticRegression(penalty='l2', solver='lbfgs', max_iter=10000, class_weight='balanced', random_state=42))\n",
    "])\n",
    "\n",
    "# Logistic Regression hyperparameter tuning\n",
    "param_grid_lr = {\n",
    "    'clf__C': [0.01, 0.1, 1]\n",
    "}\n",
    "\n",
    "lr_search = RandomizedSearchCV(lr_pipeline, param_grid_lr, n_iter=3, scoring='f1_weighted', n_jobs=-1, cv=skf)\n",
    "lr_search.fit(X_train, y_train)\n",
    "print(\"\\nBest LR params:\", lr_search.best_params_)\n",
    "y_pred_lr = lr_search.best_estimator_.predict(X_test)\n",
    "\n",
    "print(\"\\nLogistic Regression Report:\")\n",
    "\n",
    "\n",
    "report = classification_report(y_test, y_pred_lr, zero_division=0)\n",
    "print(report)\n",
    "\n",
    "# Save to a text file\n",
    "with open(\"/trinity/home/atheodosiadou/macro/results/ChIP_comb/models/fibrosis/H3K27ac_EGR1_LPS/classification_report_lr_merged_manorm_008.txt\", \"w\") as f:\n",
    "    f.write(report)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# XGBoost pipeline\n",
    "xgb_pipeline = ImbPipeline([\n",
    "    ('smote', SMOTE(random_state=42)),\n",
    "    ('clf', XGBClassifier(eval_metric='mlogloss', random_state=42))\n",
    "])\n",
    "\n",
    "# XGBoost hyperparameter tuning\n",
    "param_grid_xgb = {\n",
    "    'clf__n_estimators': [100, 200, 500],\n",
    "    'clf__max_depth': [3, 5, 7],\n",
    "    'clf__learning_rate': [0.01, 0.1, 0.2],\n",
    "    'clf__gamma': [0, 0.1, 0.2]\n",
    "}\n",
    "\n",
    "xgb_search = RandomizedSearchCV(xgb_pipeline, param_grid_xgb,  n_iter=5, scoring='f1_weighted', n_jobs=-1, cv=skf)\n",
    "xgb_search.fit(X_train, y_train)\n",
    "print(\"\\nBest XGB params:\", xgb_search.best_params_)\n",
    "y_pred_xgb = xgb_search.best_estimator_.predict(X_test)\n",
    "\n",
    "report = classification_report(y_test, y_pred_xgb)\n",
    "print(report)\n",
    "\n",
    "# Save to a text file\n",
    "with open(\"/trinity/home/atheodosiadou/macro/results/ChIP_comb/models/fibrosis/H3K27ac_EGR1_LPS/classification_report_xgb_merged_manorm_008.txt\", \"w\") as f:\n",
    "    f.write(report)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # SVM pipeline\n",
    "# svm_pipeline = ImbPipeline([\n",
    "#     ('smote', SMOTE(random_state=42)),\n",
    "#     ('clf', SVC(kernel='linear', class_weight='balanced', probability=False, random_state=42))\n",
    "# ])\n",
    "\n",
    "# # # SVM hyperparameter tuning\n",
    "# # param_grid_svm = {\n",
    "# #     'clf__C': [0.1, 1, 10],\n",
    "# #     'clf__gamma': [0.001, 0.01, 0.1]\n",
    "# # }\n",
    "# # svm_search = RandomizedSearchCV(svm_pipeline, param_grid_svm, n_iter=5, scoring='f1_weighted', n_jobs=-1, cv=skf)\n",
    "# # svm_search.fit(X_train, y_train)\n",
    "# # print(\"\\nBest SVM params:\", svm_search.best_params_)\n",
    "\n",
    "# # y_pred_svm = svm_search.best_estimator_.predict(X_test)\n",
    "\n",
    "\n",
    "# # svm_pipeline.fit(X_train, y_train)\n",
    "# # print(\"\\nBest SVM params:\", svm_pipeline.best_params_)\n",
    "\n",
    "# # y_pred_svm = svm_pipeline.best_estimator_.predict(X_test)\n",
    "\n",
    "\n",
    "# # svm_scores = cross_val_score(svm_pipeline, X_train, y_train, cv=skf, scoring='f1_weighted')\n",
    "# # print(\"\\nSVM CV F1 Scores:\", svm_scores)\n",
    "# svm_pipeline.fit(X_train, y_train)\n",
    "# y_pred_svm = svm_pipeline.predict(X_test)\n",
    "# print(\"\\nSVM Report:\")\n",
    "\n",
    "# report = classification_report(y_test, y_pred_svm, zero_division=0) #  Handle zero division\n",
    "# print(report)\n",
    "\n",
    "# # Save to a text file\n",
    "# with open(\"/trinity/home/atheodosiadou/macro/results/ChIP_comb/models/fibrosis/H3K27ac_EGR1_LPS/classification_report_svm_merged_manorm_008.txt\", \"w\") as f:\n",
    "#     f.write(report)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SHAP values for best RF\n",
    "sample_X = shap.sample(pd.DataFrame(X_test, columns=X.columns), 100, random_state=42)\n",
    "explainer = shap.TreeExplainer(rf_search.best_estimator_.named_steps['clf'])\n",
    "shap_values = explainer(sample_X)\n",
    "shap.summary_plot(shap_values, features=sample_X, feature_names=[sample_X.columns[i] for i in range(sample_X.shape[1])], show=False)\n",
    "plt.title(\"SHAP Summary Plot for Random Forest Model - H3K27ac_EGR1_LPS\")\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"/trinity/home/atheodosiadou/macro/results/ChIP_comb/models/fibrosis/H3K27ac_EGR1_LPS/shap_summary_merged_manorm_008.png\", dpi=150)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SHAP values for best XGB\n",
    "sample_X = shap.sample(pd.DataFrame(X_test, columns=X.columns), 100, random_state=42)\n",
    "explainer = shap.Explainer(xgb_search.best_estimator_.named_steps['clf'])\n",
    "shap_values = explainer(sample_X)\n",
    "shap.summary_plot(shap_values, features=sample_X, feature_names=[sample_X.columns[i] for i in range(sample_X.shape[1])], show=False)\n",
    "plt.title(\"SHAP Summary Plot for XGBoost Model - H3K27ac_EGR1_LPS\")\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"/trinity/home/atheodosiadou/macro/results/ChIP_comb/models/fibrosis/H3K27ac_EGR1_LPS/shap_summary_xgb_merged_manorm_008.png\", dpi=150)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Confusion Matrix\n",
    "sns.heatmap(confusion_matrix(y_test, y_pred_rf), annot=True, fmt='d', cmap='Blues')\n",
    "plt.title(\"Random Forest Confusion Matrix - H3K27ac_EGR1_LPS\")\n",
    "plt.xlabel(\"Predicted\")\n",
    "plt.ylabel(\"Actual\")\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"/trinity/home/atheodosiadou/macro/results/ChIP_comb/models/fibrosis/H3K27ac_EGR1_LPS/confusion_matrix_rf_merged_manorm_008.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Confusion Matrix\n",
    "sns.heatmap(confusion_matrix(y_test, y_pred_xgb), annot=True, fmt='d', cmap='Blues')\n",
    "plt.title(\"XGBoost Confusion Matrix - H3K27ac_EGR1_LPS\")\n",
    "plt.xlabel(\"Predicted\")\n",
    "plt.ylabel(\"Actual\")\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"/trinity/home/atheodosiadou/macro/results/ChIP_comb/models/fibrosis/H3K27ac_EGR1_LPS/confusion_matrix_xgb_merged_manorm_008.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- Append this block at the end of your script ---\n",
    "\n",
    "# == Analyzing most important peaks for mvalue = 1 and overlap = 1 ==\n",
    "\n",
    "# Recreate column names from X after preprocessing\n",
    "column_names = X.columns.tolist()\n",
    "print(column_names)\n",
    "\n",
    "# Convert X_test back to DataFrame\n",
    "X_test_df = pd.DataFrame(X_test, columns=column_names)\n",
    "# print(X_test_df.head())\n",
    "X_test_df = X_test_df.reset_index(drop=True)\n",
    "# print(X_test_df)\n",
    "y_test_reset = y_test.reset_index(drop=True)\n",
    "# print(y_test_reset)\n",
    "\n",
    "\n",
    "# Merge with original test rows for context\n",
    "original_test_data = data.iloc[y_test.index].reset_index(drop=True) #this will keep the original rows intact and the original column names\n",
    "# print(original_test_data.head())\n",
    "\n",
    "# Filter for rows where mvalue == 1 and overlap == 1\n",
    "filtered_idx = (y_test_reset == 2) & (X_test_df['overlap'] == 1) #false for the cases where mvalue == 0 or 1, and true for the cases where mvalue == 2\n",
    "# print(filtered_idx)\n",
    "X_filtered = X_test_df[filtered_idx]\n",
    "# print(X_filtered)\n",
    "original_filtered = original_test_data[filtered_idx].copy() #uses the same rows as X_filtered but keeps the original columns\n",
    "# print(original_filtered)\n",
    "\n",
    "# SHAP explainer from RF already defined: 'explainer'\n",
    "# Compute SHAP values on filtered rows\n",
    "print(\"\\nComputing SHAP values for mvalue=2 and overlap=1 subset...\")\n",
    "\n",
    "shap_values_filtered = explainer(X_filtered)\n",
    "# print(shap_values_filtered)\n",
    "shap_vals_array = shap_values_filtered.values\n",
    "# print(shap_vals_array)\n",
    "print(f\"SHAP values shape: {shap_vals_array.shape}\")\n",
    "\n",
    "# Compute mean absolute SHAP values across filtered samples\n",
    "mean_abs_shap = np.abs(shap_vals_array[:, :, 2]).mean(axis=0)\n",
    "print(mean_abs_shap)  \n",
    "shap_feature_importance = pd.DataFrame({\n",
    "    'Feature': column_names,\n",
    "    'Mean_Abs_SHAP': mean_abs_shap\n",
    "}).sort_values(by='Mean_Abs_SHAP', ascending=False)\n",
    "\n",
    "# Output top features\n",
    "print(\"\\nTop features contributing to mvalue = 2 and overlap = 1:\")\n",
    "print(shap_feature_importance.head(10))\n",
    "\n",
    "# Sample-level total contribution\n",
    "shap_total_contrib = np.sum(np.abs(shap_vals_array[:, :, 2]), axis=1)\n",
    "print(f\"Total SHAP contributions shape: {shap_total_contrib.shape}\")\n",
    "print(f\"Total SHAP contributions: {shap_total_contrib}\")\n",
    "original_filtered['Total_SHAP'] = shap_total_contrib\n",
    "\n",
    "# Sort and show top peaks\n",
    "top_peaks = original_filtered.sort_values(by='Total_SHAP', ascending=False).head(100)\n",
    "print(\"\\nTop peaks contributing to mvalue = 2 and overlap = 1:\")\n",
    "print(top_peaks[['chr_h3k', 'start_h3k', 'end_h3k', 'symbol', 'annotation_gene','fibrosis', 'Total_SHAP']])\n",
    "\n",
    "# Save to CSV\n",
    "top_peaks.to_csv(\"/trinity/home/atheodosiadou/macro/results/ChIP_comb/models/fibrosis/H3K27ac_EGR1_LPS/top_peaks_mval1_overlap1_merged_manorm_008.csv\", sep=\"\\t\",  index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # 4. Feature Selection using Permutation Importance\n",
    "# smote = SMOTE(random_state=42)\n",
    "# X_train_smote, y_train_smote = smote.fit_resample(X_train, y_train)\n",
    "# print(f\"After SMOTE, X_train shape: {X_train_smote.shape}, y_train shape: {y_train_smote.shape}\")\n",
    "# y_train_smote.hist()\n",
    "# model_for_feature_selection = RandomForestClassifier(n_estimators=1000, random_state=42)\n",
    "# model_for_feature_selection.fit(X_train_smote, y_train_smote)\n",
    "\n",
    "# # Select features based on importance\n",
    "# sfm = SelectFromModel(model_for_feature_selection, threshold=0.0001)  \n",
    "# sfm.fit(X_train_smote, y_train_smote)\n",
    "# X_train_selected = sfm.transform(X_train_smote)\n",
    "# X_test_selected = sfm.transform(X_test)\n",
    "\n",
    "# # Get selected feature names\n",
    "# selected_feature_indices = sfm.get_support(indices=True)\n",
    "# selected_feature_names = [X.columns[i] for i in selected_feature_indices]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from sklearn.model_selection import GridSearchCV\n",
    "\n",
    "# # Define parameter grid for Random Forest\n",
    "# param_grid = {\n",
    "#     'n_estimators': [500,1000],  # Number of trees in the forest\n",
    "#     'max_depth': [None, 10, 20],\n",
    "#     # 'min_samples_split': [2, 5, 10],\n",
    "#     'min_samples_leaf': [1, 2, 4],\n",
    "#     'max_features': ['sqrt', 'log2'],\n",
    "#     # 'bootstrap': [True, False],\n",
    "#     'criterion': ['gini', 'entropy'],\n",
    "#     # 'max_samples': [None, 0.5, 0.75],  # Optional, for sub-sampling\n",
    "#     # 'max_leaf_nodes': [None, 10, 20, 30],  # Optional, for limiting leaf nodes\n",
    "#     # 'min_weight_fraction_leaf': [0.1, 0.2],  # Optional, for handling class imbalance\n",
    "#     # 'ccp_alpha': [0.0, 0.01, 0.1],  # Optional, for pruning\n",
    "#     # 'class_weight': ['balanced', 'balanced_subsample']  # Optional, for handling class imbalance\n",
    "# }\n",
    "\n",
    "# # GridSearchCV\n",
    "# rf = RandomForestClassifier(random_state=42)\n",
    "# grid_search = GridSearchCV(rf, param_grid, cv=5, n_jobs=-1, scoring='f1_macro')\n",
    "# grid_search.fit(X_train_selected, y_train_smote)  # Use balanced y after SMOTE\n",
    "\n",
    "# print(\"Best parameters found:\", grid_search.best_params_)\n",
    "# print(\"Best cross-validated F1 score:\", grid_search.best_score_)\n",
    "\n",
    "# # Use the best estimator for further analysis\n",
    "# model = grid_search.best_estimator_\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # 6. Model Evaluation\n",
    "# y_pred = model.predict(X_test_selected)\n",
    "# balanced_accuracy = balanced_accuracy_score(y_test, y_pred)\n",
    "# print(f\"Balanced Accuracy: {balanced_accuracy}\")\n",
    "# # accuracy = accuracy_score(y_test, y_pred)\n",
    "# # print(f\"Accuracy: {accuracy}\")\n",
    "# print(classification_report(y_test, y_pred))\n",
    "\n",
    "# cm = confusion_matrix(y_test, y_pred)\n",
    "# sns.heatmap(cm, annot=True, fmt=\"d\", cmap=\"Blues\")\n",
    "# plt.xlabel(\"Predicted\")\n",
    "# plt.ylabel(\"Actual\")\n",
    "# plt.title(\"Confusion Matrix\")\n",
    "# plt.savefig(\"/trinity/home/atheodosiadou/macro/results/ChIP_comb/models/H3K27ac_EGR1_LPS/confusion_matrix_minority_no_start_end.png\")\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # 7. Feature Importance Plot (using selected features)\n",
    "# result = permutation_importance(model, X_test_selected, y_test, n_repeats=10, random_state=42)\n",
    "# importances = result.importances_mean\n",
    "# indices = np.argsort(importances)[::-1]\n",
    "# print(indices)\n",
    "# print(\"Selected feature names:\", selected_feature_names)\n",
    "# print(\"Feature importances (selected features):\")\n",
    "\n",
    "# plt.figure(figsize=(12, 10))\n",
    "# plt.title(\"Feature Importances (Selected Features)\")\n",
    "# plt.bar(range(X_train_selected.shape[1]), importances[indices], align=\"center\")\n",
    "# plt.xticks(range(X_train_selected.shape[1]), [selected_feature_names[i] for i in indices], rotation=90)\n",
    "# plt.xlabel(\"Feature\")\n",
    "# plt.ylabel(\"Importance\")\n",
    "# plt.tight_layout()\n",
    "# plt.savefig(\"/trinity/home/atheodosiadou/macro/results/ChIP_comb/models/H3K27ac_EGR1_LPS/feature_importances_selected_0001_minority.png\")\n",
    "# plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "macro_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
